server/src/main/java/org/elasticsearch/transport/LeakTracker.java

```diff
@@ -0,0 +1,273 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.transport;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.Randomness;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+/**
+ * Leak tracking mechanism that allows for ensuring that a resource has been properly released before a given object is garbage collected.
+ *
+ */
+public final class LeakTracker {
+
+    private static final Logger logger = LogManager.getLogger(LeakTracker.class);
+
+    private static final int TARGET_RECORDS = 25;
+
+    private final Set<Leak<?>> allLeaks = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+    private final ReferenceQueue<Object> refQueue = new ReferenceQueue<>();
+    private final ConcurrentMap<String, Boolean> reportedLeaks = ConcurrentCollections.newConcurrentMap();
+
+    public static final LeakTracker INSTANCE = new LeakTracker();
+
+    private LeakTracker() {
+    }
+
+    /**
+     * Track the given object.
+     *
+     * @param obj object to track
+     * @return leak object that must be released by a call to {@link Leak#close(Object)} before {@code obj} goes out of scope
+     */
+    public <T> Leak<T> track(T obj) {
+        reportLeak();
+        return new Leak<>(obj, refQueue, allLeaks);
+    }
+
+    private void clearRefQueue() {
+        while (true) {
+            Leak<?> ref = (Leak<?>) refQueue.poll();
+            if (ref == null) {
+                break;
+            }
+            ref.dispose();
+        }
+    }
+
+    private void reportLeak() {
+        if (logger.isErrorEnabled() == false) {
+            clearRefQueue();
+            return;
+        }
+        while (true) {
+            Leak<?> ref = (Leak<?>) refQueue.poll();
+            if (ref == null) {
+                break;
+            }
+
+            if (ref.dispose() == false) {
```

Comment:
Can we just do this and not bother with the separate `clearRefQueue()` method?

```suggestion
            if (ref.dispose() == false || logger.isErrorEnabled() == false) {
```

Reply:
++ sounds good